src/rustup-utils/src/utils.rs

```diff
@@ -121,149 +131,150 @@ pub fn canonicalize_path(path: &Path, notify_handler: NotifyHandler) -> PathBuf
 }
 
 pub fn tee_file<W: io::Write>(name: &'static str, path: &Path, w: &mut W) -> Result<()> {
-    raw::tee_file(path, w).map_err(|e| {
-        Error::ReadingFile {
+    Ok(try!(raw::tee_file(path, w).chain_err(|| {
+        ErrorKind::ReadingFile {
             name: name,
             path: PathBuf::from(path),
-            error: e,
         }
-    })
+    })))
 }
 
 pub fn download_file(url: hyper::Url,
                      path: &Path,
                      hasher: Option<&mut Hasher>,
                      notify_handler: NotifyHandler)
                      -> Result<()> {
+    use hyper::status::StatusCode::NotFound;
+
     notify_handler.call(Notification::DownloadingFile(&url, path));
-    raw::download_file(url.clone(), path, hasher, notify_handler).map_err(|e| {
-        Error::DownloadingFile {
-            url: url,
-            path: PathBuf::from(path),
-            error: e,
+    match raw::download_file(url.clone(), path, hasher, notify_handler) {
+        Ok(_) => Ok(()),
+        Err(e @ raw::DownloadError::Status(NotFound)) => {
+            Err(e).chain_err(|| ErrorKind::Download404 {
+                url: url,
+                path: path.to_path_buf(),
+            })
         }
-    })
+        Err(e) => {
+            Err(e).chain_err(|| ErrorKind::DownloadingFile {
+                url: url,
+                path: path.to_path_buf(),
+            })
+        }
+    }
 }
 
 pub fn parse_url(url: &str) -> Result<hyper::Url> {
-    hyper::Url::parse(url).map_err(|_| Error::InvalidUrl { url: url.to_owned() })
+    Ok(try!(hyper::Url::parse(url)
+            .chain_err(|| ErrorKind::InvalidUrl { url: url.to_owned() })))
 }
 
 pub fn cmd_status(name: &'static str, cmd: &mut Command) -> Result<()> {
-    raw::cmd_status(cmd).map_err(|e| {
-        Error::RunningCommand {
+    Ok(try!(raw::cmd_status(cmd).chain_err(|| {
+        ErrorKind::RunningCommand {
             name: OsString::from(name),
-            error: e,
         }
-    })
+    })))
 }
 
 pub fn assert_is_file(path: &Path) -> Result<()> {
     if !is_file(path) {
-        Err(Error::NotAFile { path: PathBuf::from(path) })
+        Err(ErrorKind::NotAFile { path: PathBuf::from(path) }.into())
     } else {
         Ok(())
     }
 }
 
 pub fn assert_is_directory(path: &Path) -> Result<()> {
     if !is_directory(path) {
-        Err(Error::NotADirectory { path: PathBuf::from(path) })
+        Err(ErrorKind::NotADirectory { path: PathBuf::from(path) }.into())
     } else {
         Ok(())
     }
 }
 
 pub fn symlink_dir(src: &Path, dest: &Path, notify_handler: NotifyHandler) -> Result<()> {
     notify_handler.call(Notification::LinkingDirectory(src, dest));
-    raw::symlink_dir(src, dest).map_err(|e| {
-        Error::LinkingDirectory {
+    Ok(try!(raw::symlink_dir(src, dest).chain_err(|| {
+        ErrorKind::LinkingDirectory {
```

Comment:
I think this pattern of wrapping in `Ok(try!` is not necessary here, an artifact of earlier iterations.